Surface flux diag

roelof-groenewald opened the pull request:
Added flux diagnostics for scraped particles. This allows us to track the current absorbed on an electrode as a function of time.

EtherealMC-Bit committed:
Merged memaster into mkrg/scraped_particle_statistics

EtherealMC-Bit committed:
Added wrapper to get number of particle species tracked by the scraper

Not sure if this is going to be useful, but it demonstrates a method to get information
  from the ParticleBoundaryBuffer into Python.

EtherealMC-Bit committed:
Test file for particle scraping

EtherealMC-Bit committed:
Merged Andrew's newest changes

EtherealMC-Bit committed:
Merge remote-tracking branch 'atmyers/particle_boundary_scrape' into mkrg/scraped_particle_statistics

EtherealMC-Bit committed:
Added wrapper to get number of particle species tracked by the scraper

Not sure if this is going to be useful, but it demonstrates a method to get information
  from the ParticleBoundaryBuffer into Python.

EtherealMC-Bit committed:
Stubbed out the main wrapper functions

EtherealMC-Bit committed:
Stubbed out the main wrapper functions

EtherealMC-Bit committed:
Added parameters to wrapper

EtherealMC-Bit committed:
Added parameters to wrapper

EtherealMC-Bit committed:
Added wrapper for getting the number of particles scraped of a species on a boundary

EtherealMC-Bit committed:
Added wrapper for getting the number of particles scraped of a species on a boundary

KZhu-ME committed:
Added timeseries class

KZhu-ME committed:
Added flux_diagnostic file

KZhu-ME committed:
Merge branch 'memaster' into kzhu/flux_diags

nruof committed:
Nruof/runinfo minimal (#44)

* Update memaster with new upstream/development (#42)

* CI: Add Missing Python Analysis for EB Test (#2147)

* CI: Add Missing Python Analysis for EB Test

* Use 1 MPI Process for Azure

* a few _rt (#2146)

* PSATD: div Cleaning Implemented only with psatd.J_linear_in_time=1 (#2142)

* BTD: Don't Flush If Written (#2148)

Written BTD buffers for lab snapshot data are reset to zero size
(count). When we do the final write of all partly filled buffers
in `FilterComputePackFlushLastTimestep`, we should not write such
already completed backtransformed lab snapshots again.

* openPMD: `groupBased` Option Missing (#2149)

The `groupBased` iteration encoding (input: `g`) was not parsed.

* Remove predefined constants from example input files (#2153)

* BTD_ReducedSliceDiag: BTD Plotfiles (#2152)

By accident, the 2nd test did not use plotfile output.

* Allow extra particle attributes (besides ux, uy, uz and w) to be set at particle creation in AddNParticles() (#2115)

* exposes AddRealComp to Python to allow extra particle attributes to be added at runtime; also includes a new function to grab a particle data array from the name of the component rather than the index

* added functionality to AddNParticles() to allow extra particle attributes to also be set during particle creation

* added function to get index of a particle component given the PID name

* changed new get component index and get_particle_arrays_from_comp_name functions to take species name as argument rather than species id

* changed warpx_addRealComp to accept a species name as input and only add the new component for that species

* added a test of the pywarpx bridge to get particle data and add new particle attributes at runtime

* changed all particle interacting functions in libwarpx to use the species name rather than id, also changed the functions to get particle array data to use the component name rather than index

* updated test according to PR #2119 changes

* removed unneeded BL_ASSERT(nattr == 1) statement

* fixed bug in add_particles to correctly determine the number of extra attributes

* fixed bug in AddNParticles if fewer attribute values are passed than the number of extra arrays for the species

* use isinstance(attr, ndarray) rather than type(attr) is np.ndarray

* generalize_runtime_comps_io

* fix OpenPMD

* fix OpenPMD

* fix plot flags in WritePlotFile

* fix offset and comment

* changed extra pid test to not use an underscore in the pid name

* switched _libwarpx.py::add_particles to use kwargs to accept the weight and extra attribute arrays

* License update in test file

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

* fix typo

* added a test with unique_particles=False

* Apply suggestions from code review

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

* updated docstring and comments

Co-authored-by: atmyers <atmyers2@gmail.com>
Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

* updated mewarpx with recent changes to _libwarpx.add_particles

* fix broken test - test_mepicmi.py

Co-authored-by: Edoardo Zoni <59625522+EZoni@users.noreply.github.com>
Co-authored-by: MaxThevenet <maxence.thevenet@desy.de>
Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>
Co-authored-by: Neïl Zaim <49716072+NeilZaim@users.noreply.github.com>
Co-authored-by: atmyers <atmyers2@gmail.com>

* Add runinfo from warp to warpx

* Doc string edit

Co-authored-by: Roelof Groenewald <40245517+roelof-groenewald@users.noreply.github.com>
Co-authored-by: Edoardo Zoni <59625522+EZoni@users.noreply.github.com>
Co-authored-by: MaxThevenet <maxence.thevenet@desy.de>
Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>
Co-authored-by: Neïl Zaim <49716072+NeilZaim@users.noreply.github.com>
Co-authored-by: atmyers <atmyers2@gmail.com>

KZhu-ME committed:
Ported more flux diags from warp

KZhu-ME committed:
Merge branch 'kzhu/flux_diags' of https://github.com/ModernElectron/WarpX into kzhu/flux_diags

KZhu-ME committed:
Renamed FluxDiag to FluxDiagnostic

roelof-groenewald committed:
added picmi arguments to scrape particles at the domain boundary

roelof-groenewald committed:
added picmi arguments to scrape particles at the domain boundary

nruof committed:
Remove js, species conversion in runinfo

EtherealMC-Bit committed:
Added wrapper to get the full particle buffer into python

EtherealMC-Bit committed:
Added wrapper to get the full particle buffer into python

roelof-groenewald committed:
rearanged the getBuffer properties code a little

roelof-groenewald committed:
rearanged the getBuffer properties code a little

KZhu-ME committed:
Added test for Timeseries

KZhu-ME committed:
Added test file for Timeseries

EtherealMC-Bit committed:
Added docstrings +other suggested changes

EtherealMC-Bit committed:
Added docstrings +other suggested changes

KZhu-ME committed:
Added appendable array

KZhu-ME committed:
Added appendable array

EtherealMC-Bit committed:
Added num_particles_impacted_boundary docstring

EtherealMC-Bit committed:
Added num_particles_impacted_boundary docstring

KZhu-ME committed:
Merge branch 'kzhu/flux_diags' of https://github.com/ModernElectron/WarpX into kzhu/flux_diags

EtherealMC-Bit committed:
fixed mistake in docstring

EtherealMC-Bit committed:
fixed mistake in docstring

EtherealMC-Bit committed:
Changed boundary parameter to be a string for clarity

EtherealMC-Bit committed:
Changed boundary parameter to be a string for clarity

EtherealMC-Bit committed:
Fixed issue with the boundary parameter for scraping

EtherealMC-Bit committed:
Fixed issue with the boundary parameter for scraping

KZhu-ME committed:
Added parallel_util and added test for flux diag

roelof-groenewald committed:
initial commit of scraping diagnostic functionality added to assemblies.py - does not work yet

roelof-groenewald committed:
Merge remote-tracking branch 'public/kzhu/flux_diags' into surfaceFluxDiag

EtherealMC-Bit committed:
Fixed issue with the boundary input for scraping stats wrapper

EtherealMC-Bit committed:
Added demonstration of particle scraping wrapper

EtherealMC-Bit committed:
Added analysis.py file

EtherealMC-Bit committed:
Fix typo in one of the dimension maps

Co-authored-by: Roelof Groenewald <40245517+roelof-groenewald@users.noreply.github.com>

roelof-groenewald committed:
changes that should be included in upstream PR as well

roelof-groenewald committed:
partly working implementation of SurfaceFluxDiag

EtherealMC-Bit committed:
Added before esolve to warpx evolve

PhilMiller committed:
Merge Andrew Myers's changes on which we're based

Merge commit '48e07504c969acd82fda2660a1583f30aa4d8c9f' into scratch

roelof-groenewald committed:
added test for the scraped particle buffer wrappers

roelof-groenewald committed:
added a comment for a needed future improvement

EtherealMC-Bit commented on mewarpx/mewarpx/appendablearray.py:
```diff
@@ -0,0 +1,242 @@
+        assert self._unitshape is not None,\
+               'Only an array with a specified unitshape can be reshaped'
+        assert len(newunitshape) == len(self._unitshape),\
+               ('New unitshape must have the same number of dimensions as original'+
+                'unitshape')
```
```suggestion
               ('New unitshape must have the same number of dimensions as original '+
                'unitshape')
```

roelof-groenewald replied:
Nice catch. The appendable array is coming from PR #47, please add this comment there.

EtherealMC-Bit commented on mewarpx/mewarpx/assemblies.py:
```diff
@@ -32,6 +48,117 @@ def getvoltage(self):
+                empty = True
+            else:
+                # get the timesteps at which particles were scraped
+                comp_steps = _libwarpx.get_particle_boundary_buffer(
+                    species.name, self.scraper_label, "stepScraped", mwxrun.lev
+                )
```
Won't the step at which particles were scraped always be included when we get the buffer from WarpX? Because the step at which they were scraped will be an attribute of the `ParticleBuffer`.

roelof-groenewald replied:
Yes that sounds correct. So do you see something in the code that should be changed due to that fact? I guess the way it is set up we do check that the buffer isn't empty before trying to access it, is that what you are referring to?

EtherealMC-Bit replied:
At the very least there's no reason to pass in the `"stepScraped"` parameter.

roelof-groenewald replied:
I don't follow. The `_libwarpx.get_particle_boundary_buffer` function requires a PID name for which you want the buffer values.

EtherealMC-Bit replied:
Never mind then

EtherealMC-Bit commented on mewarpx/mewarpx/runinfo.py:
```diff
@@ -0,0 +1,479 @@
+        self._save_version_info()
+        self._save_comp_info()
+
+    def _save_version_info(self):
+        """Save info about file versions."""
+        # self.run_param_dict['metools_version'] = metools.__version__
```
```suggestion
    def _save_version_info(self):
        """Save info about file versions."""
        raise NotImplementedError
        # self.run_param_dict['metools_version'] = metools.__version__
```
I think moving `NotImplementedErrors` to before commented out code is cleaner, Peter has had me do that before.

roelof-groenewald replied:
Also a comment for the other PR.

EtherealMC-Bit commented on mewarpx/mewarpx/runinfo.py:
```diff
@@ -0,0 +1,479 @@
+        raise NotImplementedError
+
+    def _save_comp_info(self):
+        """Save info about the computer."""
+        # self.run_param_dict['uname'] = platform.uname()
```
```suggestion
        """Save info about the computer."""
        raise NotImplementedError
        # self.run_param_dict['uname'] = platform.uname()
```

roelof-groenewald replied:
This comment can also go to PR #47.

EtherealMC-Bit reviewed (changes requested):
This is really cool, just some minor changes.

atmyers committed:
Add support for EB buffer

atmyers committed:
protect for AMREX_USE_EB, static_assert that EB and RZ aren't both on.

atmyers committed:
fix unused

atmyers committed:
add inputs file

atmyers committed:
add test

atmyers committed:
fix bugs, remove print

atmyers committed:
fix test

atmyers committed:
fix test path.

atmyers committed:
remove no-op code

atmyers committed:
adding clear particles method

atmyers committed:
attempt at adding time stamp

EtherealMC-Bit committed:
Moved python PICMI particle boundary scrape test

EtherealMC-Bit committed:
Merged in Andrew's most recent changes

Including:
- Scraping from embedded boundaries
- A test for scraping
- A method for clearing particles
- The first attempt at adding a time step to scraped particles

PhilMiller commented on Source/Diagnostics/FlushFormats/FlushFormat.H:
```diff
@@ -5,6 +5,7 @@
 
 #include "Diagnostics/ParticleDiag/ParticleDiag.H"
 #include "Particles/MultiParticleContainer.H"
+#include "WarpX.H"
```
Is this needed for something referenced in the header? I don't see anything added. This can potentially cause rebuilds for lots of irrelevant changes.

roelof-groenewald replied:
This change is from Andrew's PR, so it would be better to just make the same comment on his upstream PR.

EtherealMC-Bit committed:
Readded python test files

EtherealMC-Bit committed:
Renamed test file to the correct name

EtherealMC-Bit committed:
Renamed test file to the correct name

EtherealMC-Bit committed:
Removed old test

KZhu-ME commented on Examples/Modules/ParticleBoundaryScrape/inputs_scrape:
```diff
@@ -0,0 +1,48 @@
+electrons.momentum_distribution_type = "constant"
+electrons.uz = 2000. # uth the std of the (unitless) momentum
+electrons.scrape_xhi = 1
+electrons.scrape_eb = 1
```
```suggestion
electrons.scrape_eb = 1

```
Just eof newline

KZhu-ME commented on Python/pywarpx/_libwarpx.py:
```diff
@@ -557,6 +559,64 @@ def get_particle_arrays(species_name, comp_name, level):
+    ----------
+
+        species_name   : the species name that the data will be returned for.
+        boundary       : the boundary from which to get the scraped particle data.
+                         In the form x/y/z_hi/lo
+        comp_name      : the component of the array data that will be returned.
+        level          : Which AMR level to retrieve scraped particle data from.
+    Returns
```
```suggestion
        level          : Which AMR level to retrieve scraped particle data from.
        
    Returns
```
I think autodoc is really picky about formatting stuff like this.

roelof-groenewald committed:
added special functionality to get the timestep at which particles were scraped

roelof-groenewald committed:
added special functionality to get the timestep at which particles were scraped

KZhu-ME commented on Python/pywarpx/_libwarpx.py:
```diff
@@ -1709,6 +1769,41 @@ def get_mesh_charge_density_fp_lovects(level, include_ghosts=True):
+
+    '''
+    if geometry_dim == '3d':
+        dimensions = {'x' : 0, 'y' : 1, 'z' : 2}
```
[Earlier in `get_particle_boundary_buffer` on line 589,](https://github.com/ModernElectron/WarpX/pull/48/files#diff-45fe8ad0cfda5ed498793502822a96a08d19326c06d5a73b5f876f37c0568364R589) the dimensions are set as `{'x' : 1, 'y' : 2, 'z' : 3}` wheres here, the dimensions are set as `{'x' : 0, 'y' : 1, 'z' : 2}`. Is this intentional?

EtherealMC-Bit replied:
Yeah that was an error. It should be 0, 1, 2.

KZhu-ME reviewed (changes requested):
just a few changes and a question.

EtherealMC-Bit committed:
removed debug print

roelof-groenewald committed:
added python wrapper for the clearParticles() function of the scraper buffer

roelof-groenewald committed:
added python wrapper for the clearParticles() function of the scraper buffer

roelof-groenewald committed:
Merge remote-tracking branch 'atmyers/particle_boundary_scrape' into surfaceFluxDiag

roelof-groenewald committed:
added special wrapper function to get the timesteps at which the particles in the boundary buffer were scraped

roelof-groenewald committed:
added special wrapper function to get the timesteps at which the particles in the boundary buffer were scraped

roelof-groenewald committed:
Merge remote-tracking branch 'public/memaster' into surfaceFluxDiag

roelof-groenewald committed:
adapted the record_scrapedparticles function to work with data from multiple tiles

roelof-groenewald committed:
Merge remote-tracking branch 'atmyers/particle_boundary_scrape' into pm/scraped-particles-buffer

roelof-groenewald committed:
updated test to match the non-PICMI test for the particle scraper buffer

roelof-groenewald committed:
Merge remote-tracking branch 'public/pm/scraped-particles-buffer' into surfaceFluxDiag

KZhu-ME commented on Python/pywarpx/_libwarpx.py:
```diff
@@ -557,6 +561,165 @@ def get_particle_arrays(species_name, comp_name, level):
+
+    '''
+    if geometry_dim == '3d':
+        dimensions = {'x' : 1, 'y' : 2, 'z' : 3}
```
Should this be `{'x' : 1, 'y' : 2, 'z' : 3}`?

KZhu-ME commented on Python/pywarpx/_libwarpx.py:
```diff
@@ -557,6 +561,165 @@ def get_particle_arrays(species_name, comp_name, level):
+    return boundary_num
+
+
+def get_particle_boundary_buffer(species_name, boundary, comp_name, level):
```
Is this a duplicate function? If not, should it be named something else so it doesn't collide with the function 2 functions above?

roelof-groenewald replied:
Yeah it is a duplicate. Thanks for catching it! The issues from multiple merges...

KZhu-ME commented on Python/pywarpx/_libwarpx.py:
```diff
@@ -557,6 +561,165 @@ def get_particle_arrays(species_name, comp_name, level):
+
+    '''
+    if geometry_dim == '3d':
+        dimensions = {'x' : 1, 'y' : 2, 'z' : 3}
```
Should this be `{'x' : 0, 'y' : 1, 'z' : 2}`?

```suggestion
        dimensions = {'x' : 0, 'y' : 1, 'z' : 2}
```

roelof-groenewald replied:
Yeap, thanks! Hopefully this will now be fixed once and for all.

EtherealMC-Bit replied:
This is fixed in an upstream commit that pulls this logic into another function to remove the code duplication, so I'll just cherry pick that here.

EtherealMC-Bit committed:
Fixed formatting mistake in docstring

KZhu-ME reviewed (changes requested):
Just a few minor things.

roelof-groenewald committed:
removed duplicate get_particle_boundary_buffer function

roelof-groenewald committed:
Merge branch 'surfaceFluxDiag' of github.com:ModernElectron/WarpX into surfaceFluxDiag

nruof commented on mewarpx/mewarpx/diags_store/flux_diagnostic.py:
```diff
@@ -167,6 +167,60 @@ def _collect_dataframe(self, clear=True):
+            surface (mewarpx.Assembly object): The assembly object in which
+                particles are scraped.
+            write_dir (string): Directory to write CSV files to.
+            kwargs: See :class:`metools.diags_store.diag_base.WarpDiagnostic`
+                for more timing options.
```
```suggestion
            kwargs: See :class:`mewarpx.diags_store.diag_base.WarpXDiagnostic`
                for more timing options.
```

EtherealMC-Bit committed:
Fixed 'metools' reference in docstring

Co-authored-by: Nicholas William Ruof <86024944+nruof@users.noreply.github.com>

EtherealMC-Bit committed:
Start to mewarp flux diag accuracy test

EtherealMC-Bit committed:
Progress on flux diag accuracy test

EtherealMC-Bit committed:
Renamed some metools to mewarpx

roelof-groenewald committed:
quick fix for runinfo RecursionError

EtherealMC-Bit committed:
Finished porting test and moved it into test_flux_diagnostic.py

EtherealMC-Bit committed:
Merge branch 'memaster' into surfaceFluxDiag

roelof-groenewald commented on mewarpx/mewarpx/assemblies.py:
```diff
@@ -32,6 +48,116 @@ def getvoltage(self):
+        Note:
+            Assumes the fixed form of fields given in Assembly().  Doesn't
+            check since this is called many times.
+        """
```
```suggestion

        Note:
            The total charge scraped and energy of the particles scraped 
            is multiplied by -1 since these quantities are leaving the system.
        """
```

roelof-groenewald commented on mewarpx/tests/test_particle_scraping_stats.py:
```diff
@@ -0,0 +1,95 @@
+"""
```
This file can now be removed, right?

EtherealMC-Bit replied:
It hasn't been copied anywhere else, so do we not need a test for this?

roelof-groenewald replied:
The same functionality is tested in _WarpX/Examples/Modules/ParticleBoundaryScrape/PICMI_inputs_scrape.py_ so we are good.

EtherealMC-Bit committed:
Add note about energy and charge multiplied by -1 for scraped particles

Co-authored-by: Roelof Groenewald <40245517+roelof-groenewald@users.noreply.github.com>

EtherealMC-Bit committed:
Removed whitespace

EtherealMC-Bit committed:
Add newline to docstring in _libwarpx.get_particle_boundary_buffer

EtherealMC-Bit committed:
Remove test_particle_scraping_stats

KZhu-ME reviewed (approved):
Looks good!

KZhu-ME committed:
Changed mcc_wrapper to add E_total pid to all species

roelof-groenewald committed:
Merge branch 'memaster' into surfaceFluxDiag

peterscherpelz commented on Python/pywarpx/Bucket.py:
```diff
@@ -68,3 +68,5 @@ def attrlist(self):
             result += [attrstring]
         return result
 
+    def __setstate__(self, state):
+        pass
```
What is this for? It at least needs an explanatory statement. It also seems like this would completely break any attempt at pickling.

roelof-groenewald replied:
Without it `runinfo` doesn't work due to an issue with recursion depth - we had this issue before with Warp as well and resolved it through the same approach. I guess it depends on what you want to pickle as to whether it will break. I believe this will break some parts of `mwxrun.simulation` if we tried to pickle it but until we find something broken by it, I'd say we should keep it in since it allows `FluxDiagFromFile` to work (specifically allows `runinfo` to be loaded). I will add the comment explanation though.

peterscherpelz commented on Source/Python/WarpXWrappers.cpp:
```diff
@@ -494,6 +495,72 @@ extern "C"
+        auto& particle_buffers = WarpX::GetInstance().GetParticleBoundaryBuffer();
+        auto& particle_buffer = particle_buffers.getParticleBuffer(species_name, boundary);
+
+        const int comp = particle_buffer.NumIntComps() - 1;
```
Assuming this is always the last integer component seems pretty brittle.

Would it make more sense to make this something like `warpx_getParticleBoundaryBufferInt` with a flexible component name? It would mirror the functionality of real components that way...

roelof-groenewald replied:
Unfortunately since the particle buffer is a `amrex::AmrParticleContainer` and not a `warpx::ParticleContainer` it doesn't support named PIDs. However, this specific integer component is added to the particle buffer after all the other PIDs have been copied over (from whatever `warpx::ParticleContainer` the buffer gets its particles) so this 'last integer component' approach can't be broken by the user at runtime. That being said I agree this should be made more durable. The best approach is probably to save the index of this component for each particle buffer in an array or vector when it is created. I'll make a new card for that on Trello and we can make that change in an upstream PR.

roelof-groenewald replied:
Alternatively, you can bring up the issue in Andrew's PR and maybe he'll just fix it before that PR is merged.

roelof-groenewald replied:
There is actually an extended problem here where any runtime attribute added to a species' particle container after the buffer for that particle (for a given boundary) has been created won't be included in the buffer. So far I don't think there is a specific case where this will be a problem for us since we generally add all the extra PID's before the first step of the simulation.

peterscherpelz commented on mewarpx/mewarpx/diags_store/flux_diagnostic.py:
```diff
@@ -845,6 +900,83 @@ def __init__(self, df, area, step_begin=None, step_end=None, dt=None):
+    def __init__(self, basedir='diags', fluxdatafile=None,
+                 fluxdatafileformat='fluxes/fluxdata*', fs=None):
+        """Load from fluxdata_XXXXXXXXXX.dpkl and runinfo.dpkl files.
+        Arguments:
```
```suggestion

        Arguments:
```

peterscherpelz commented on mewarpx/mewarpx/diags_store/flux_diagnostic.py:
```diff
@@ -845,6 +900,83 @@ def __init__(self, df, area, step_begin=None, step_end=None, dt=None):
+                 fluxdatafileformat='fluxes/fluxdata*', fs=None):
+        """Load from fluxdata_XXXXXXXXXX.dpkl and runinfo.dpkl files.
+        Arguments:
+            basedir (str): Base directory of the diagnostic files.
+                Has to contain a runinfo dill-pickled file.
```
Should note that by default will search `diags` and `diags_*` directories...

peterscherpelz commented on mewarpx/mewarpx/diags_store/flux_diagnostic.py:
```diff
@@ -845,6 +900,83 @@ def __init__(self, df, area, step_begin=None, step_end=None, dt=None):
+        self.__dict__.update(dict_to_load)
+
+    def _get_runinfofile(self, basedir):
+        """Function to look for a runinfo.dpkl file in the current directory."""
```
I don't really understand the use of this function and the comment is not correct. In addition to needing a better comment, this will fail if `runinfo.dpkl` is present but no flux diag files were found, with a very misleading error.

Possible approach: Make a list of directories = ['diags'] + reversed(sorted(glob command stuff)). Iterate through them to find first one that contains both runinfo and flux diags. If it fails, print out an error message that no directory was found with both files.

roelof-groenewald replied:
Actually we don't need any of that since we don't have restart in WarpX that will rename 'diags' so I just removed all the extra functionality. If we implement that sort of diagnostic handling in WarpX restarts it'll be easy to just reference the `metools` version of this function to regain the removed functionality.

peterscherpelz commented on mewarpx/mewarpx/setups_store/diode_setup.py:
```diff
@@ -434,6 +434,14 @@ def init_conductors(self):
+            self.ions.scrape_zmin = 1
+            self.ions.scrape_zmax = 1
+
     def init_scraper(self):
```
Do we still need this function then?

EtherealMC-Bit replied:
Looks like there's an input parameter for picmi species to set this called `warpx_save_particles_at_blank` where blank is one of `xlo`, `xhi`, `zlo`, `zhi`, `ylo`, `yhi`, or `eb`. So these values can be passed in when the species is initialized.

But how about we add another input parameter for diode setup which is a list of strings for which boundaries to scrape? Then we use those to set all those `warpx_save_particles_at_blank` parameters.

roelof-groenewald replied:
We could reorganize things such that the `init_scraper` function will set the `scrape_zmin` and `scrape_zmax` values for each species. This is probably a better approach since if we activate the scraping but don't install a flux diagnostic the particle buffer will continue to grow indefinitely. So the best is probably to have the `init_fluxdiag` function call the `init_scraper` function.

EtherealMC-Bit replied:
So should I add the input parameter I mentioned so that we can choose what boundaries to scrape from or just keep it hardcoded to xlo/hi like it is now?

roelof-groenewald replied:
@MKieburtz Since in a diode we will always have exactly a cathode and anode there is no need to add the option of scraping in the x-direction. But please do make the change to move lines 437-443 to be inside the `init_scraper` function and add a call to that function to the `init_fluxdiag` function. You can then also remove the `init_scraper` flag.

peterscherpelz commented on Source/WarpX.H:
```diff
@@ -556,6 +558,7 @@ public:
     void setplot_rho (bool a_plot_rho) {plot_rho = a_plot_rho;}
 
     int maxStep () const {return max_step;}
+
```
I'd avoid newlines like this that'll put us out of sync with upstream. Same in particle boundary buffer .cpp and _libwarpx.py

peterscherpelz reviewed (changes requested):
Thanks! Mostly looks good and wonderful to have this functionality in. We may need discussion on what to do with int components - see comments.

Finding runinfo in FluxDiagFromFile and the bucket setstate are both key issues to resolve/improve.

EtherealMC-Bit committed:
Cleanup from code review

- Removed some newlines that aren't upstream
- Removed some unused imports in the flux diag test
- Added a newline in a docstring

EtherealMC-Bit committed:
Move code for enabling scraping into init_scraper

- Also remove the init_scraper flag because init_scraper() is called from
  init_conductors()

EtherealMC-Bit committed:
removed test_calc_dt, which I accidently added

roelof-groenewald committed:
fixed issue with printing currents for each conductor during flux diag output and other requested change from PR review

peterscherpelz reviewed (approved):
Thanks all!
